servicetalk-data-jackson/src/main/java/io/servicetalk/data/jackson/JacksonStreamingSerializer.java

```diff
@@ -143,53 +164,33 @@ boolean consumeOnNext(final Buffer buffer) throws IOException {
-                                                    final ByteBufferFeeder feeder) {
-                super(subscriber, reader, parser);
-                this.feeder = feeder;
             }
 
-            @Override
-            boolean consumeOnNext(final Buffer buffer) throws IOException {
+            /**
+             * Feeds the buffer into the {@link ByteBufferFeeder}.
+             * @param buffer the buffer to feed into the streaming JSON parser.
+             * @param feeder the feeder into which the buffer should be fed.
+             * @throws IOException if feeding the buffer failed.
+             */
+            private void feedByteBufferParser(final Buffer buffer, final ByteBufferFeeder feeder) throws IOException {
                 feeder.feedInput(buffer.toNioBuffer());
-                return feeder.needMoreInput();
-            }
-        }
-
-        private abstract static class DeserializeSubscriber<T> implements Subscriber<Buffer> {
-            private final JsonParser parser;
-            private final ObjectReader reader;
-            private final Deque<JsonNode> tokenStack = new ArrayDeque<>(8);
-            private final Subscriber<? super Iterable<T>> subscriber;
-            @Nullable
-            private Subscription subscription;
-            @Nullable
-            private String fieldName;
-
-            private DeserializeSubscriber(final Subscriber<? super Iterable<T>> subscriber,
-                                          final ObjectReader reader,
-                                          final JsonParser parser) {
-                this.reader = reader;
-                this.parser = parser;
-                this.subscriber = subscriber;
             }
 
             /**
-             * Consumer the buffer from {@link #onNext(Buffer)}.
-             * @param buffer The bytes to append.
-             * @return {@code true} if more data is required to parse an object. {@code false} if object(s) should be
-             * parsed after this method returns.
-             * @throws IOException If an exception occurs while appending {@link Buffer}.
+             * Initializes the JSON streaming parser and feeder based on the buffer type.
+             * @param arrayBackedBuffer if the buffer on the operator is backed by an array.
+             * @throws IOException if creating the non-blocking {@link JsonParser} failed.
              */
-            abstract boolean consumeOnNext(Buffer buffer) throws IOException;
+            private void initParserAndFeeder(final boolean arrayBackedBuffer) throws IOException {
+                if (arrayBackedBuffer) {
+                    parser = reader.getFactory().createNonBlockingByteArrayParser();
+                    isByteArrayParser = true;
+                } else {
+                    parser = reader.getFactory().createNonBlockingByteBufferParser();
+                    isByteArrayParser = false;
```

Comment:
If `ByteBufferFeeder` is available now, I wonder if we can simplify the control flow to always use it and forget about ByteArrayParser. The `ByteBuffer` can be backed by direct memory or heap memory. I would expect it to work with any type of `ByteBuffer`. WDYT?

Reply:
For backwards compatibility (e.g. if users link an older version of Jackson) we may have to still fallback to `ByteArray`. This is assuming perf is roughly equivalent (I expect some additional object allocation overhead if `byte[]` needs to be wrapped in ByteBuffer, and I'm not sure how efficient our `toNioBuffer()` is ..).